src/lib/logstore/log_dev.hpp

```diff
@@ -838,73 +758,84 @@ class LogDev : public std::enable_shared_from_this< LogDev > {
     void do_load(off_t offset);
+    void assert_next_pages(log_stream_reader& lstream);
 
-#if 0
-    log_group_header* read_validate_header(uint8_t* buf, uint32_t size, bool* read_more);
-    sisl::byte_array read_next_header(uint32_t max_buf_reads);
-#endif
+    /// @brief force to flush the log device
+    /// @return whether real flush is done
+    bool flush();
 
-    void _persist_info_block();
-    void assert_next_pages(log_stream_reader& lstream);
-    void set_flush_status(bool flush_status);
-    bool get_flush_status();
+    bool can_flush_in_this_thread();
 
 private:
-    std::unique_ptr< sisl::StreamTracker< log_record > >
-        m_log_records;                              // The container which stores all in-memory log records
-    std::atomic< logid_t > m_log_idx{0};            // Generator of log idx
-    std::atomic< int64_t > m_pending_flush_size{0}; // How much flushable logs are pending
-    std::atomic< bool > m_is_flushing{false}; // Is LogDev currently flushing (so far supports one flusher at a time)
-    bool m_stopped{true}; // Is Logdev stopped. We don't need lock here, because it is updated under flush lock
+    std::unique_ptr< sisl::StreamTracker< log_record > > m_log_records; // Container stores all in-memory log records
+    std::atomic< logid_t > m_log_idx{0};                                // Generator of log idx
+    std::atomic< int64_t > m_pending_flush_size{0};                     // How much flushable logs are pending
+    bool m_stopped{false}; // Is Logdev stopped. We don't need lock here, because it is updated under flush lock
     logdev_id_t m_logdev_id;
-    JournalVirtualDev* m_vdev{nullptr};
+    std::shared_ptr< JournalVirtualDev > m_vdev;
     shared< JournalVirtualDev::Descriptor > m_vdev_jd; // Journal descriptor.
     HomeStoreSafePtr m_hs;                             // Back pointer to homestore
+    flush_mode_t m_flush_mode;
 
     folly::SharedMutexWritePriority m_store_map_mtx;
     std::unordered_map< logstore_id_t, logstore_info > m_id_logstore_map;
     std::unordered_map< logstore_id_t, uint64_t > m_unopened_store_io;
     std::unordered_set< logstore_id_t > m_unopened_store_id;
-    std::unordered_map< logstore_id_t, logid_t > m_last_flush_info;
-
     std::multimap< logid_t, logstore_id_t > m_garbage_store_ids;
     Clock::time_point m_last_flush_time;
 
     logid_t m_last_flush_idx{-1}; // Track last flushed, last device offset and truncated log idx
-    off_t m_last_flush_dev_offset{0};
-    logid_t m_last_truncate_idx{-1};
+    logid_t m_last_truncate_idx{std::numeric_limits< logid_t >::min()}; // logdev truncate up to this idx
+
+    // after completing flush, we can not truncate the log records directly. since we might shedule new flush/append in
+    // the callback of the append_async, which can be found in the logstore test case.
+    // this means we might schedule call on_flush_completion in another on_flush_completion, which is a recursive call.
```

Comment:
now we still under the `flush guard`  in `on_flush_completion`,  how can  we have recursive call to `on_flush_completion`?

Reply:
logstore supports inline flush mode.  in this flush mode , when we use append_async and in the callback we schedule another flush or append_async , we need to do flush before the previous flush is completed, since the callback will be called in `on_write_completion` , which is also under the protection of flush guard.
pls refer to https://github.com/eBay/HomeStore/blob/39781c3fb7e7b4da3f77f0a9ebb3e0ad73f2ddfc/src/tests/test_log_store.cpp#L153

Reply:
sorry if this is a stupid question.
Dont we get deadlock if we are under `flush guard` and calling ` append_async` which is trying to obtain the `flush guard` again?

Reply:
1 if this two happens in different thread, `append_async` will try to acquire the flush lock, , it will not block if it can not acquire the lock.

2 if this two happens in the same thread, since we use `std:recursive_mutex` now, which can be acquired for multiple times , it will not block and not dead lock happen